### DISCUSSION TITLE
Get data block layout

### DISCUSSION THREAD

**Discussion by lucapiccio** [gijzelaerr/python-snap7#498] [Q&A] [answered]
Hi, i'm building a program to transfer S7 Db to SQL (mysql / sqlite). It's done and it works.

It's possible to retrive the db data structure programmaticaly? (pointers and type data)

I need to create a flexible system, if technicien change a type of data or add any variable to db i need to rebuild the structure of communication and db.

I haven't saw anything on the doc for this.

**Comment by lupaulus** [comment]
Hey @lucapiccio, In TIA Portal it is possible to export the DB DataStructure, within a Text file (.db) with the structure of the DataBlock. Problem there aren't any information of the offset address. In Simatic Manager, there are possibility to export the DB in XML (PLCTags.xml) with the WinCC TagConverter (Offset and type included)

**Reply by lupaulus** [reply]
https://youtu.be/WfGRIGWz9jg?si=l-m82a9vcd6PPN_C

**Comment by lucapiccio** [comment]
Hi, thank's for the reply, but i know well Tia and S7 and this is not a problem. 
My question is if exist a mode (exemple SZL query) to retrive this information programmaticaly.

The idea is write a convertor from S7 to Mysql / sqlite3 with automatic regeneration of tables programmaticaly based on retrived DB Layout.

But the problem is that i don't found any doc about how to query SZL (read_szl / read_szl_list) or other system to get DB headers pointers.

**Reply by lupaulus** [reply]
Hey,

I doesn't know if it is possible directly from the S7 Protokol (with the SZL Headers), I have no experience with them.
As mentioned, it is possible to do it by generating the files described in my last reply.

Is a Read_SZL functionality implemented in the snap-S7 lib ?

**Reply by lucapiccio** [reply]
Seems implemented but not documented.

            snap7/client/__init__.py:    def read_szl_list(self) -> bytearray
            snap7/client/__init__.py:    def read_szl(self, ssl_id: int, index: int = 0x0000) -> S7SZL
            snap7/types.py:class S7SZL(ctypes.Structure)

I'm trying to read and effectly it return somthing:

            client = snap7.client.Client()
            #p = client.read_szl_list()
            
            p = client.read_szl(0,0)
            h = p.Header
            for field in h._fields_:
                header = field[0]
                headerVal = getattr(h, field[0])
                print(h,header,headerVal)
            print(p.Data)

return:

            <S7SZLHeader LengthDR: 8, NDR: 113> LengthDR 8
            <S7SZLHeader LengthDR: 8, NDR: 113> NDR 113
            <snap7.types.c_byte_Array_16380 object at 0x7efef366d130>

From s7 siemens documentation:
SZL_ID (W#16#...)

            0111 -> identification
            0012 -> category
            0F12 -> partial headers 
            0113 -> 1 record on the memory zone pointed
            0113 + index 1 -> working memory
            .....

But i don't know how arrange data readed in S7SZL(ctypes.Structure) and i do not understand why it ask un ssl_id as argoument and non un SZL_ID (W#16#...) and how to convert this

Thanks for any suggestion

**Reply by lupaulus** [reply]
I will give some tries next week !
Links about your S7 Siemens documentation ?

**Comment by lucapiccio** [comment]
Hi
https://support.industry.siemens.com/cs/mdm/109755202?c=22058958475&lc=en-AT

**Comment by lupaulus** [comment]
@nikteliy https://github.com/gijzelaerr/python-snap7/issues/196, seems you have written this functionality, maybe you can help ?

**Comment by lupaulus** [comment]
SSL (English) = SZL (German)

![image](https://github.com/gijzelaerr/python-snap7/assets/20111917/0e04fed4-66da-4147-b4a1-22885f3a8f3e)

**Reply by luewh** [reply]
![image](https://github.com/gijzelaerr/python-snap7/assets/85931693/4d3c585d-48b5-40e4-907f-ef757307e10d)
(tested on s7-1500 with ssl_id=0x0000, index=0x0000)
That how Wireshark understands the S7SZL.Data.

![image](https://github.com/gijzelaerr/python-snap7/assets/85931693/172105f3-653f-4798-9f33-2b0a52b9327a)
(tested on s7-1500 with ssl_id=0x0011, index=0x0000)

list of possible ssl_id:
ssl_id = [0x0000, 0x0011, 0x001c, 0x0024, 0x0074, 0x0124, 0x0424, 0x0524, 0x0f00, 0x0f11, 0x0f1c, 0x0f24, 0x0f74]

hope it will help you all, I'm also interested in getting data block layout "programmably".

**Reply by lupaulus** [reply]
@luewh I will add this to documentation as soon as possible on my side

**Comment by gijzelaerr** [comment, marked as answer]
Closing as answered. DB structure/layout information is not available via the S7 protocol. TIA Portal export or XML export tools are the correct approach.
